File: backend/app/modules/documents/pdf.py
"""PDF access: per-page classification, rendering, and native word extraction.

PyMuPDF is imported lazily inside functions so the package remains importable in
environments without it. This module never converts coordinates — it reports raw
point/pixel geometry plus the page dimensions, and `extraction/ir.py` normalises.
"""
from __future__ import annotations

from dataclasses import dataclass, field

from app.core.config import settings
from app.core.logging import get_logger
from app.modules.documents.rectify import auto_rectify
from app.schemas.common import PageClassification

log = get_logger(__name__)


@dataclass(frozen=True)
class NativeWord:
    text: str
    x1: float
    y1: float
    x2: float
    y2: float
    block_no: int
    line_no: int


@dataclass
class PageRender:
    page_number: int
    width: float           # original page width, in points
    height: float          # original page height, in points
    dpi: int
    image_bytes: bytes
    image_width: int
    image_height: int
    classification: PageClassification
    native_words: list[NativeWord] = field(default_factory=list)


def images_to_pdf(images: list[bytes]) -> bytes:
    """Wrap one or more raw image files (JPEG/PNG) into a single PDF, one page per
    image in the given order, each page sized to that image's own pixel dimensions.

    Lets a photographed question paper or answer sheet -- one photo per page, the
    normal shape of a student's upload -- enter the exact same downstream pipeline
    (render_pages, auto_rectify, classification) as a real PDF; nothing below this
    needs to know the source wasn't one. The page's absolute size in points is
    otherwise meaningless here (render_pages re-rasterises at its own configured
    DPI), so pixel count doubling as point count is fine -- only the aspect ratio,
    which this preserves exactly, matters.
    """
    import pymupdf

    document = pymupdf.open()
    for image_bytes in images:
        pixmap = pymupdf.Pixmap(image_bytes)
        page = document.new_page(width=pixmap.width, height=pixmap.height)
        page.insert_image(page.rect, stream=image_bytes)
    data = document.tobytes()
    document.close()
    return data


def _open(data: bytes):
    import pymupdf

    return pymupdf.open(stream=data, filetype="pdf")


def classify_page(page, threshold: float | None = None) -> PageClassification:
    """Per-page classification (ADR-005): text coverage decides the extraction path."""
    threshold = settings.searchable_coverage_threshold if threshold is None else threshold
    page_area = float(page.rect.width * page.rect.height)
    if page_area <= 0:
        return PageClassification.IMAGE
    text_area = 0.0
    has_text = False
    for word in page.get_text("words"):
        x1, y1, x2, y2, text = word[0], word[1], word[2], word[3], word[4]
        if not str(text).strip():
            continue
        has_text = True
        text_area += max(0.0, x2 - x1) * max(0.0, y2 - y1)
    if not has_text:
        # No glyphs at all: an image-only page if it carries images, otherwise a blank scan.
        return PageClassification.IMAGE if page.get_images(full=True) else PageClassification.SCANNED
    return (
        PageClassification.SEARCHABLE
        if text_area / page_area > threshold
        else PageClassification.SCANNED
    )


def render_pages(data: bytes, dpi: int | None = None) -> list[PageRender]:
    """Render every page and classify it. Pages are rendered even when searchable,
    because the UI highlights against a page image regardless of extraction method."""
    dpi = dpi or settings.render_dpi
    document = _open(data)
    renders: list[PageRender] = []
    try:
        for index in range(document.page_count):
            page = document[index]
            classification = classify_page(page)
            zoom = dpi / 72.0
            long_edge_px = max(page.rect.width, page.rect.height) * zoom
            if long_edge_px > settings.render_max_long_edge:
                zoom *= settings.render_max_long_edge / long_edge_px
            import pymupdf

            pixmap = page.get_pixmap(matrix=pymupdf.Matrix(zoom, zoom), alpha=False)
            words = _native_words(page)

            image_bytes = pixmap.tobytes("png")
            width, height = float(page.rect.width), float(page.rect.height)
            image_width, image_height = pixmap.width, pixmap.height

            # A native PDF page is already a perfect rectangle -- only a
            # photographed page (no vector text layer) can need de-skewing this
            # way. Silently keeps the original image when no confident page
            # boundary is found: this step must never be required to succeed.
            if settings.auto_rectify_photos and classification != PageClassification.SEARCHABLE:
                rectified = auto_rectify(image_bytes)
                if rectified is not None:
                    log.info(
                        "page auto-rectified",
                        extra={"page": index + 1, "width": rectified.width, "height": rectified.height},
                    )
                    image_bytes = rectified.image_bytes
                    image_width, image_height = rectified.width, rectified.height
                    # No vector "points" concept applies to a photo: the rectified
                    # image's own pixel grid becomes the page frame every bbox is
                    # normalised against (extraction/ir.py), so width/height just
                    # track it 1:1 rather than the PDF's arbitrary page-rect size.
                    width, height = float(image_width), float(image_height)

            renders.append(
                PageRender(
                    page_number=index + 1,
                    width=width,
                    height=height,
                    dpi=int(round(72.0 * zoom)),
                    image_bytes=image_bytes,
                    image_width=image_width,
                    image_height=image_height,
                    classification=classification,
                    native_words=words,
                )
            )
    finally:
        document.close()
    return renders


def _native_words(page) -> list[NativeWord]:
    """Native words in DISPLAY space — the same frame as `page.rect` and the pixmap.

    PyMuPDF reports text geometry in the page's UNROTATED space, while `page.rect` and
    `get_pixmap()` are both post-rotation. On a `/Rotate 90` scan the two frames are
    transposed, so normalising a raw word box against `page.rect` puts the highlight on
    a different part of the page entirely (and can push a coordinate out of [0,1], where
    clamping hides it). `page.rotation_matrix` is the map between the two frames; it is
    the identity when the page is not rotated.
    """
    import pymupdf

    matrix = page.rotation_matrix
    out: list[NativeWord] = []
    for w in page.get_text("words"):
        if not str(w[4]).strip():
            continue
        box = pymupdf.Rect(w[0], w[1], w[2], w[3]) * matrix
        box.normalize()
        out.append(
            NativeWord(
                text=str(w[4]),
                x1=float(box.x0), y1=float(box.y0), x2=float(box.x1), y2=float(box.y1),
                block_no=int(w[5]), line_no=int(w[6]),
            )
        )
    return out


def group_words_into_lines(words: list[NativeWord]) -> list[tuple[str, tuple[float, float, float, float]]]:
    """Group native words into lines using PyMuPDF's own block/line indices."""
    buckets: dict[tuple[int, int], list[NativeWord]] = {}
    for word in words:
        buckets.setdefault((word.block_no, word.line_no), []).append(word)
    lines: list[tuple[str, tuple[float, float, float, float]]] = []
    for key in sorted(buckets):
        group = sorted(buckets[key], key=lambda w: w.x1)
        text = " ".join(w.text for w in group).strip()
        if not text:
            continue
        box = (
            min(w.x1 for w in group),
            min(w.y1 for w in group),
            max(w.x2 for w in group),
            max(w.y2 for w in group),
        )
        lines.append((text, box))
    return lines

File: apps/web/lib/server/modules/documents/pdf.ts
/**
 * PDF access: per-page classification, rendering, and native word extraction.
 * Port of backend/app/modules/documents/pdf.py, replacing PyMuPDF with
 * `pdfjs-dist` (pure JS/WASM) + `@napi-rs/canvas` (prebuilt native binaries, safe on
 * Vercel's Node serverless runtime — unlike `canvas`/`node-canvas`, which needs a
 * native compile step Vercel's build image does not reliably provide).
 *
 * This module never converts coordinates — it reports raw geometry plus page
 * dimensions, and extraction/geometry.ts normalises. Perspective-rectification of
 * photographed pages (rectify.py) is deferred out of Phase 1 (see the migration
 * plan): Gemini vision tolerates skew/lighting far better than a local OCR engine
 * did, so the value of that step is much lower now than it was.
 */
import { createCanvas } from "@napi-rs/canvas";
import { PDFDocument as PdfLibDocument } from "pdf-lib";
import * as pdfjsLib from "pdfjs-dist/legacy/build/pdf.mjs";
import type { PDFPageProxy, PageViewport } from "pdfjs-dist";

/** pdfjs-dist's `TextItem` type isn't re-exported from the package root (only from an
 * internal path), so this mirrors the handful of fields actually used here. */
interface TextItem {
  str: string;
  transform: number[];
  width: number;
  hasEOL: boolean;
}
import { settings } from "@/lib/server/config";
import { CorruptDocumentError, EncryptedPdfError, TooManyPagesError } from "@/lib/server/errors";
import type { PageClassification } from "@/lib/server/modules/extraction/types";

// pdfjs-dist's legacy build (ESM, since v4 dropped the CJS build) works without a
// browser DOM as long as it is given a canvas factory; there's no worker here since
// we call it directly, in-process, one Vercel function invocation at a time.

export interface NativeWord {
  text: string;
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  lineIndex: number;
}

export interface PageRender {
  pageNumber: number;
  width: number; // original page width, in PDF points (scale=1 viewport)
  height: number; // original page height, in PDF points
  dpi: number;
  imageBytes: Buffer;
  imageWidth: number;
  imageHeight: number;
  classification: PageClassification;
  nativeWords: NativeWord[];
}

class NodeCanvasFactory {
  create(width: number, height: number) {
    const canvas = createCanvas(width, height);
    const context = canvas.getContext("2d");
    return { canvas, context };
  }
  reset(canvasAndContext: { canvas: unknown }, width: number, height: number) {
    const canvas = canvasAndContext.canvas as ReturnType<typeof createCanvas>;
    canvas.width = width;
    canvas.height = height;
  }
  destroy(_canvasAndContext: unknown) {
    // no-op: GC handles it
  }
}

async function openDocument(data: Buffer) {
  // `canvasFactory` is a real, documented pdfjs-dist runtime option (its own Node.js
  // examples pass it) that the published v4 types don't declare — hence the cast.
  const params = {
    data: new Uint8Array(data),
    canvasFactory: new NodeCanvasFactory(),
    isEvalSupported: false,
    useSystemFonts: true,
  } as unknown as Parameters<typeof pdfjsLib.getDocument>[0];
  const loadingTask = pdfjsLib.getDocument(params);
  try {
    return await loadingTask.promise;
  } catch (err) {
    const name = (err as { name?: string })?.name ?? "";
    if (name === "PasswordException") throw new EncryptedPdfError();
    throw new CorruptDocumentError(String((err as Error)?.message ?? err));
  }
}

/** Opens just far enough to report page count / encryption, for upload validation. */
export async function inspectPdf(data: Buffer): Promise<{ pageCount: number }> {
  const doc = await openDocument(data);
  try {
    const pageCount = doc.numPages;
    if (pageCount === 0) throw new CorruptDocumentError("The PDF contains no pages.");
    if (pageCount > settings.maxPages) {
      throw new TooManyPagesError(`The document exceeds the page cap.`, {
        pageCount,
        cap: settings.maxPages,
      });
    }
    return { pageCount };
  } finally {
    await doc.destroy();
  }
}

/** Wraps one or more raw JPEG/PNG images into a single PDF, one page per image, each
 * page sized to that image's own pixel dimensions — lets a photographed sheet enter
 * the same downstream pipeline as a real PDF. */
export async function imagesToPdf(images: Buffer[]): Promise<Buffer> {
  const doc = await PdfLibDocument.create();
  for (const bytes of images) {
    const isPng = bytes[0] === 0x89 && bytes[1] === 0x50;
    const image = isPng ? await doc.embedPng(bytes) : await doc.embedJpg(bytes);
    const page = doc.addPage([image.width, image.height]);
    page.drawImage(image, { x: 0, y: 0, width: image.width, height: image.height });
  }
  return Buffer.from(await doc.save());
}

const SEARCHABLE_COVERAGE = () => settings.searchableCoverageThreshold;

/** Item transform -> viewport-space rect. pdf.js gives glyph transforms in PDF user
 * space; `Util.transform(viewport.transform, item.transform)` composes them into the
 * same top-left-origin, y-down frame as the rendered raster (this is the same
 * technique pdf.js's own text-layer builder uses). Ascent/descent are approximated as
 * fixed fractions of the font-transform height, since exact font metrics aren't
 * available here — fine for a line-grouping/highlight box, not meant to be
 * typographically exact. */
function itemRect(item: TextItem, viewport: PageViewport) {
  const tx = pdfjsLib.Util.transform(viewport.transform, item.transform);
  const fontHeight = Math.hypot(tx[2], tx[3]);
  const widthScale = Math.hypot(tx[0], tx[1]) || 1;
  const width = item.width * widthScale;
  const x = tx[4];
  const yBaseline = tx[5];
  const y1 = yBaseline - fontHeight * 0.8;
  const y2 = yBaseline + fontHeight * 0.25;
  return { x1: x, y1, x2: x + width, y2 };
}

async function extractNativeWords(page: PDFPageProxy, viewport: PageViewport): Promise<NativeWord[]> {
  const content = await page.getTextContent();
  const words: NativeWord[] = [];
  let lineIndex = 0;
  for (const item of content.items) {
    if (!("str" in item)) continue; // skip TextMarkedContent entries
    const text = item.str;
    if (text && text.trim()) {
      const rect = itemRect(item, viewport);
      words.push({ text, ...rect, lineIndex });
    }
    if (item.hasEOL) lineIndex += 1;
  }
  return words;
}

/** Text coverage ratio decides the extraction path (ADR-005). */
async function classifyPage(page: PDFPageProxy, viewport: PageViewport): Promise<PageClassification> {
  const pageArea = viewport.width * viewport.height;
  if (pageArea <= 0) return "scanned";
  const words = await extractNativeWords(page, viewport);
  if (words.length === 0) return "scanned";
  const textArea = words.reduce(
    (sum, w) => sum + Math.max(0, w.x2 - w.x1) * Math.max(0, w.y2 - w.y1),
    0
  );
  return textArea / pageArea > SEARCHABLE_COVERAGE() ? "searchable" : "scanned";
}

export async function renderPages(data: Buffer, dpi?: number): Promise<PageRender[]> {
  const targetDpi = dpi ?? settings.renderDpi;
  const doc = await openDocument(data);
  const renders: PageRender[] = [];
  try {
    for (let index = 0; index < doc.numPages; index++) {
      const page: PDFPageProxy = await doc.getPage(index + 1);
      const baseViewport = page.getViewport({ scale: 1 });
      const classification = await classifyPage(page, baseViewport);
      const nativeWords = await extractNativeWords(page, baseViewport);

      let zoom = targetDpi / 72;
      const longEdgePx = Math.max(baseViewport.width, baseViewport.height) * zoom;
      if (longEdgePx > settings.renderMaxLongEdge) {
        zoom *= settings.renderMaxLongEdge / longEdgePx;
      }
      const viewport = page.getViewport({ scale: zoom });
      const canvasFactory = new NodeCanvasFactory();
      const { canvas, context } = canvasFactory.create(
        Math.ceil(viewport.width),
        Math.ceil(viewport.height)
      );
      // @napi-rs/canvas's context is API-compatible with the DOM CanvasRenderingContext2D
      // pdf.js's browser-oriented types expect; the cast bridges the two type identities.
      await page.render({ canvasContext: context as unknown as CanvasRenderingContext2D, viewport }).promise;
      const imageBytes = canvas.toBuffer("image/png");

      renders.push({
        pageNumber: index + 1,
        width: baseViewport.width,
        height: baseViewport.height,
        dpi: Math.round(72 * zoom),
        imageBytes,
        imageWidth: canvas.width,
        imageHeight: canvas.height,
        classification,
        nativeWords,
      });
    }
  } finally {
    await doc.destroy();
  }
  return renders;
}

/** Groups native words into lines, by the line index pdf.js's own text stream
 * already assigns via `hasEOL` — the flattened equivalent of PyMuPDF's
 * (block_no, line_no) grouping, since pdf.js doesn't expose block structure
 * directly. */
export function groupWordsIntoLines(words: NativeWord[]): Array<[string, [number, number, number, number]]> {
  const buckets = new Map<number, NativeWord[]>();
  for (const word of words) {
    const bucket = buckets.get(word.lineIndex) ?? [];
    bucket.push(word);
    buckets.set(word.lineIndex, bucket);
  }
  const lines: Array<[string, [number, number, number, number]]> = [];
  for (const key of [...buckets.keys()].sort((a, b) => a - b)) {
    const group = [...buckets.get(key)!].sort((a, b) => a.x1 - b.x1);
    const text = group
      .map((w) => w.text)
      .join(" ")
      .trim();
    if (!text) continue;
    const box: [number, number, number, number] = [
      Math.min(...group.map((w) => w.x1)),
      Math.min(...group.map((w) => w.y1)),
      Math.max(...group.map((w) => w.x2)),
      Math.max(...group.map((w) => w.y2)),
    ];
    lines.push([text, box]);
  }
  return lines;
}
